### DISCUSSION TITLE
Open alpha when all buffers closed?

### DISCUSSION THREAD

**Discussion by bennypowers** [goolord/alpha-nvim#85] [Q&A] [answered]
Hello!

Using `nvim-neo-tree` and `bufferline.nvim`, I'd like for nvim-alpha to load up whenever I close all the open, non-tree buffers.

Currently, If i close all those buffers, I get an empty `[No Name]` buffer, I'd like to see alpha instead.

How?

**Comment by GallowsDove** [comment]
Also tried to do this, however it seems that trying almost any possible autocmd with alpha.start() failed with a lot of errors.

**Comment by Bekaboo** [comment]
I've tried `vim.cmd [[ au BufDelete * if empty(filter(tabpagebuflist(), '!buflisted(v:val)')) | Alpha | endif ]]`, which works for Startify, but for Alpha I get `Error detected while processing BufDelete Autocommands for "*":` and `E488: Trailing characters:  Alpha | endif`.

**Reply by GallowsDove** [reply]
You can fix that by using exec "Alpha" instead of just alpha. But like I've said, it still threw errors for me

**Reply by Bekaboo** [reply]
Thanks!
I find this works for me (nvim 0.7.0-dev):
```
-- Automatically open alpha when the last buffer is deleted and only one window left
vim.cmd [[ au BufDelete * if empty(filter(tabpagebuflist(), '!buflisted(v:val)')) && winnr('$') == 1 | exec 'Alpha' | endif ]]
```

**Reply by GallowsDove** [reply]
Unfortunately that throws error when there is more than one tab and one is closed for me.

```
Error detected while processing function <lambda>651[1]..<SNR>104__set_buf:                                                        
line    5:                                                                                                                         
E5555: API call: Vim(lua):E5108: Error executing lua .../name/.local/share/nvim/plugged/alpha-nvim/lua/alpha.lua:556: Cursor positi
on outside buffer                                                                                                                  
stack traceback:                                                                                                                   
        [C]: in function 'nvim_win_set_cursor'                                                                                     
        .../name/.local/share/nvim/plugged/alpha-nvim/lua/alpha.lua:556: in function 'redraw'                                      
        [string ":lua"]:1: in main chunk                                                                                           
Error detected while processing function <lambda>696[1]..<SNR>104__set_buf:                                                        
line    5:                                                                                                                         
E5555: API call: Vim(lua):E5108: Error executing lua .../name/.local/share/nvim/plugged/alpha-nvim/lua/alpha.lua:556: Cursor positi
on outside buffer                                                                                                                  
stack traceback:                                                                                                                   
        [C]: in function 'nvim_win_set_cursor'                                                                                     
        .../name/.local/share/nvim/plugged/alpha-nvim/lua/alpha.lua:556: in function 'redraw'                                      
        [string ":lua"]:1: in main chunk                                                                                           
```

**Comment by bennypowers** [comment, marked as answer]
~~I worked this out in a roundabout way by binding a specific callback to delete buffers (and as the callback to delete buffers in bufferline.nvm)~~

~~Perhaps bufdelete.nvim will add a User event or callback option, in which case this would be a drop easier to configure. See https://github.com/famiu/bufdelete.nvim/issues/18~~

famiu kindly added `BDeletePre` to bufdelete.nvim, so this is good:


```lua
---@author kikito
---@see https://codereview.stackexchange.com/questions/268130/get-list-of-buffers-from-current-neovim-instance
local function get_listed_buffers()
  local buffers = {}
  local len = 0
  for buffer = 1, vim.fn.bufnr('$') do
    if vim.fn.buflisted(buffer) == 1 then
      len = len + 1
      buffers[len] = buffer
    end
  end

  return buffers
end

vim.api.nvim_create_augroup('alpha_on_empty', { clear = true })
vim.api.nvim_create_autocmd('User', {
  pattern = 'BDeletePre',
  group = 'alpha_on_empty',
  callback = function(event)
    local found_non_empty_buffer = false
    local buffers = get_listed_buffers()

    for _, bufnr in ipairs(buffers) do
      if not found_non_empty_buffer then
        local name = vim.api.nvim_buf_get_name(bufnr)
        local ft = vim.api.nvim_buf_get_option(bufnr, 'filetype')

        if bufnr ~= event.buf and name ~= '' and ft ~= 'Alpha' then
          found_non_empty_buffer = true
        end
      end
    end

    if not found_non_empty_buffer then
      require 'neo-tree'.close_all()
      vim.cmd [[:Alpha]]
    end
  end,
})

```

**Reply by AmetistDrake** [reply]
This is not working anymore, because `pattern = 'BDeletePre'` [now](https://github.com/famiu/bufdelete.nvim/tree/0b9c0f182b09c51170bb9f252f15de7695e4d507) also contains the bufnr in the pattern.

Here is an up to date solution: 

```
vim.api.nvim_create_augroup("alpha_on_empty", { clear = true })
vim.api.nvim_create_autocmd("User", {
	pattern = "BDeletePre *",
	group = "alpha_on_empty",
	callback = function()
		local bufnr = vim.api.nvim_get_current_buf()
		local name = vim.api.nvim_buf_get_name(bufnr)

		if name == "" then
      vim.cmd([[:Alpha | bd#]])
		end
	end,
})
```

**Comment by mtarld** [comment]
You can leverage the `BDeletePost` event as well:
```lua
vim.api.nvim_create_autocmd("User", {
  pattern = "BDeletePost",
  callback = function(event)
    local fallback_name = vim.api.nvim_buf_get_name(event.buf)
    local fallback_ft = vim.api.nvim_buf_get_option(event.buf, "filetype")
    local fallback_on_empty = fallback_name == "" and fallback_ft == ""

    if fallback_on_empty then
      -- do you magic there
    end
  end,
})
```

**Comment by latipun7** [comment]
With this breaking change commit (https://github.com/famiu/bufdelete.nvim/commit/bdaae05ccdc9abf975599e8db9acec6b8ef271b2), pattern need to be `BDeletePre*` or `BDeletePost*` for those code above to works.

**Comment by Allaman** [comment]
Does this work for you correctly? 
This is my autocmd. 
```lua
vim.api.nvim_create_augroup("alpha_on_empty", { clear = true })
vim.api.nvim_create_autocmd("User", {
  pattern = "BDeletePost*",
  group = "alpha_on_empty",
  callback = function(event)
    local fallback_name = vim.api.nvim_buf_get_name(event.buf)
    local fallback_ft = vim.api.nvim_buf_get_option(event.buf, "filetype")
    local fallback_on_empty = fallback_name == "" and fallback_ft == ""

    if fallback_on_empty then
      vim.cmd("Alpha")
    end
  end,
})
```
Somehow a "[No Name]" buffer is created which messes up the handling of opening/deleting multiple buffers in sequence

**Reply by latipun7** [reply]
Yes it works correctly for me. Make sure you are using latest commit of bufdelete.nvim

**Reply by Allaman** [reply]
I am on the latest and greatest plugin versions ;)

To be sure can you please test the following sequence:
- Open two files in tow buffers
- Close both buffers
- Alpha is back
- Again open two files
- Close one file
- Now, for me here Alpha appears again although there is an opened buffer

Thank you!

**Reply by latipun7** [reply]
Sorry for late reply. I also has created [No Name] buffer, but it still has correct behavior for me.

https://user-images.githubusercontent.com/20012970/195380597-67bcdeb1-a87d-4b08-a2c2-4579242c7a88.mp4

So, I updated my autocommand to this below for correct behavior that I want:
```lua
  -- need bufdelete.nvim, neo-tree & alpha-dashboard
  local alpha_on_empty = create_augroup("alpha_on_empty", { clear = true })
  create_aucmd("User", {
    pattern = "BDeletePost*",
    group = alpha_on_empty,
    callback = function(event)
      local fallback_name = vim.api.nvim_buf_get_name(event.buf)
      local fallback_ft = vim.api.nvim_buf_get_option(event.buf, "filetype")
      local fallback_on_empty = fallback_name == "" and fallback_ft == ""

      if fallback_on_empty then
        require("neo-tree").close_all()
        vim.cmd("Alpha")
        vim.cmd(event.buf .. "bwipeout")
      end
    end,
  })
```

https://user-images.githubusercontent.com/20012970/195382005-0dbf8fa7-c509-4f25-a694-5d32efc71d8f.mp4

**Reply by jarmos-san** [reply]
Thanks for sharing a working code snippet but I stumbled a bit of a caveat with your version of the code. So, if `neo-tree` is lazy-loaded, Neovim will complain with a `"neo-tree" module not found` error message.

So, I wrapped the module import in a `pcall()` function. Here's my updated version of the same code:

```lua
vim.api.nvim_create_autocmd("User", {

  -- stuff....

  callback = function()
    if fallback_on_empty then
      local ok, _ = pcall(require, "neo-tree")

      if not ok then
        vim.cmd("packadd neo-tree")
        -- Use the latest recommended approach to handle Neotree. See the docs for info:
        -- https://github.com/nvim-neo-tree/neo-tree.nvim/blob/ab8ca9fac52949d7a741b538c5d9c3898cd0f45a/doc/neo-tree.txt#L140
        vim.cmd("Neotree close")
      end

      vim.cmd("Alpha")
      vim.cmd(event.buf .. "bwipeout")
    end
  end,
})
```

**Comment by Allaman** [comment]
Thanks for your effort! The only difference was that you were opening files via neo-tree and I via Telescope. But with your updated autocmd both work flawless for me as well!

**Reply by majamin** [reply]
Doesn't work for me in my full setup, so I created a minimal config. What am I doing wrong?
https://gist.github.com/majamin/374717acfb1d190c79177269d139b005

**Reply by Allaman** [reply]
Your minimal setup works for me. Do you close a buffer with correct commands from the bufdelete package like `:Bdelete` ?

**Reply by majamin** [reply]
Thank you @Allaman that's what it was (closing buffer with plugin `Bdelete` works). I will try this on my full config.

**Reply by majamin** [reply]
No conflicts in my full config as well 👍

**Comment by JirR02** [comment]
So I know I'm a bit late to the party but since bufdelete.nvim is archived, what are the alternatives? I mean sure it will work at the moment but I don't know for how long.

Would this just imply replacing `BDeletePre *` with `BufDelete` from the [bufferline.nvim](https://github.com/akinsho/bufferline.nvim) plugin?

**Reply by Allaman** [reply]
[Snacks bufdelete](https://github.com/folke/snacks.nvim/blob/main/docs/bufdelete.md)?

**Reply by jarmos-san** [reply]
[echasnovski/mini.bufremove](https://github.com/echasnovski/mini.bufremove)
